ibis/backends/pandas/executor/windows.py

```diff
@@ -0,0 +1,173 @@
+from __future__ import annotations
+
+import pandas as pd
+
+import ibis.expr.operations as ops
+from ibis.backends.pandas.executor.core import execute
+from ibis.backends.pandas.executor.utils import asseries
+from ibis.util import gen_name
+
+
+class UngroupedFrame:
+    def __init__(self, df):
+        self.df = df
+
+    def groups(self):
+        yield self.df
+
+    def apply_reduction(self, func, **kwargs):
+        result = func(self.df, **kwargs)
+        data = [result] * len(self.df)
+        return pd.Series(data, index=self.df.index)
+
+    def apply_analytic(self, func, **kwargs):
+        return func(self.df, **kwargs)
+
+
+class GroupedFrame:
+    def __init__(self, df, group_keys):
+        self.df = df
+        self.group_keys = group_keys
+        self.groupby = df.groupby(group_keys, as_index=True)
+
+    def groups(self):
+        for _, df in self.groupby:
+            yield df
+
+    def apply_analytic(self, func, **kwargs):
+        results = [func(df, **kwargs) for df in self.groups()]
+        return pd.concat(results)
+
+    def apply_reduction(self, func, **kwargs):
+        name = gen_name("result")
+        result = self.groupby.apply(func, **kwargs).rename(name)
+        df = self.df.merge(result, left_on=self.group_keys, right_index=True)
+        return df[name]
+
+
+class RowsFrame:
+    def __init__(self, parent):
+        self.parent = parent
+
+    @staticmethod
+    def adjust(length, index, start_offset, end_offset):
+        if start_offset is None:
+            start_index = 0
+        else:
+            start_index = index + start_offset
+            if start_index < 0:
+                start_index = 0
+            elif start_index > length:
+                start_index = length
+
+        if end_offset is None:
+            end_index = length
+        else:
+            end_index = index + end_offset + 1
+            if end_index < 0:
+                end_index = 0
+            elif end_index > length:
+                end_index = length
+
+        return (start_index, end_index)
+
+    def apply_analytic(self, func, **kwargs):
+        return self.parent.apply_analytic(func, **kwargs)
+
+    def apply_reduction(self, func, **kwargs):
```

Comment:
There are interesting and more optimal ways to compute this, but we can leave that for a follow-up (or never!)